zcash_primitives/src/transaction/mod.rs

```diff
@@ -498,16 +632,185 @@ impl Transaction {
+                    "invalid consensus branch id: ".to_owned() + e,
+                )
+            })
+        })?;
+        let lock_time = reader.read_u32::<LittleEndian>()?;
+        let expiry_height: BlockHeight = reader.read_u32::<LittleEndian>()?.into();
+        Ok((consensus_branch_id, lock_time, expiry_height))
+    }
 
+    #[allow(clippy::redundant_closure)]
+    fn read_v5_sapling<R: Read>(
+        mut reader: R,
+    ) -> io::Result<Option<sapling::Bundle<sapling::Authorized>>> {
+        let sd_v5s = Vector::read(&mut reader, SpendDescriptionV5::read)?;
+        let od_v5s = Vector::read(&mut reader, OutputDescriptionV5::read)?;
+        let n_spends = sd_v5s.len();
+        let n_outputs = od_v5s.len();
+        let value_balance = if n_spends > 0 || n_outputs > 0 {
+            Self::read_amount(&mut reader)?
+        } else {
+            Amount::zero()
+        };
+
+        let anchor = if n_spends > 0 {
+            Some(sapling::read_base(&mut reader, "anchor")?)
+        } else {
+            None
+        };
+
+        let v_spend_proofs = Array::read(&mut reader, n_spends, |r| sapling::read_zkproof(r))?;
+        let v_spend_auth_sigs = Array::read(&mut reader, n_spends, |r| {
+            SpendDescription::read_spend_auth_sig(r)
+        })?;
+        let v_output_proofs = Array::read(&mut reader, n_outputs, |r| sapling::read_zkproof(r))?;
+
+        let binding_sig = if n_spends > 0 || n_outputs > 0 {
+            Some(redjubjub::Signature::read(&mut reader)?)
+        } else {
+            None
+        };
+
+        let shielded_spends = sd_v5s
+            .into_iter()
+            .zip(
+                v_spend_proofs
+                    .into_iter()
+                    .zip(v_spend_auth_sigs.into_iter()),
+            )
+            .map(|(sd_5, (zkproof, spend_auth_sig))| {
+                // the following `unwrap` is safe because we know n_spends > 0.
+                sd_5.into_spend_description(anchor.unwrap(), zkproof, spend_auth_sig)
+            })
+            .collect();
+
+        let shielded_outputs = od_v5s
+            .into_iter()
+            .zip(v_output_proofs.into_iter())
+            .map(|(od_5, zkproof)| od_5.into_output_description(zkproof))
+            .collect();
+
+        Ok(binding_sig.map(|binding_sig| sapling::Bundle {
+            value_balance,
+            shielded_spends,
+            shielded_outputs,
+            authorization: sapling::Authorized { binding_sig },
+        }))
+    }
+
+    fn read_v5_orchard<R: Read>(
+        mut reader: R,
+    ) -> io::Result<Option<orchard::Bundle<orchard::bundle::Authorized, Amount>>> {
+        let actions_without_auth = Vector::read(&mut reader, |r| {
+            orchard_serialization::read_action_without_auth(r)
+        })?;
+        if actions_without_auth.is_empty() {
+            Ok(None)
+        } else {
+            let flags = orchard_serialization::read_flags(&mut reader)?;
```

Comment:
The implementation of `read_flags` enforces that unused bits are zero. This is as intended but was ambiguous in the spec. The spec has been clarified in version 2021.2.4.